aten/src/ATen/native/quantized/cpu/qnnpack_utils.h

```diff
@@ -221,3 +227,69 @@ Tensor qnnpack_avg_pool2d(
 } // namespace native
 } // namespace at
 #endif
+
+namespace {
+std::vector<float> generate_requantization_scales(
+    const at::Tensor& weight_scales,
+    const float input_scale,
+    const float output_scale) {
+  // Since weight scale is allocated with padding
+  // weight_scales.numel() gives us padded num elements.
+  auto num_output_channels_padded = weight_scales.numel();
+  float* weight_scales_data = weight_scales.data_ptr<float>();
+  std::vector<float> requant_scales(num_output_channels_padded, 1.f);
+  for (int i = 0; i < num_output_channels_padded; ++i) {
+    requant_scales[i] = weight_scales_data[i] * input_scale / output_scale;
+  }
+  return requant_scales;
+}
+
+std::pair<at::Tensor, at::Tensor> make_zero_points_and_scales_tensor(
+    const at::Tensor& weight_contig
+    ) {
+  auto num_output_channels = weight_contig.size(0);
+  // Add 8 to account for bufferring needed by QNNPACK.
+  auto num_output_channels_padded = weight_contig.size(0) + 8;
+  const auto qtype = weight_contig.qscheme();
+  at::Tensor weight_zp =
+    at::_empty_affine_quantized(
+        {1},
+        at::device(at::kCPU).dtype(at::kQUInt8),
+        1.f, 0);
+  // Adjust weight zero point, similar to weight data.
+  uint8_t* weight_zp_data;
+  if (qtype == at::kPerTensorAffine) {
+    weight_zp_data = (uint8_t*)weight_zp.data_ptr<c10::quint8>();
+    weight_zp_data[0] = (uint8_t)(weight_contig.q_zero_point() + 128);
+  } else if (qtype == at::kPerChannelAffine) {
+    weight_zp.resize_({num_output_channels_padded});
```

Comment:
Maybe zero-init the values after `num_output_channels` just to make this deterministic.

Reply:
Sounds good.